cli/containerdutils/hostpaths.go

```diff
@@ -0,0 +1,294 @@
+package containerdutils
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/net/context"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/cio"
+	"github.com/docker/docker/pkg/stringid"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+// VerifyHostDirExists ensures that the desired directory exists on the host
+// nolint: gocyclo
+func VerifyHostDirExists(ctx context.Context, client WrappedContainerdClient, dirname string, image containerd.Image) error {
+	//logrus.Infof("XXX VerifyHostDirExists %s", dirname)
+	testerSpec := specs.Spec{
+		Root: &specs.Root{
+			Path: "rootfs",
+		},
+		Process: &specs.Process{
+			Cwd: "/",
+			Args: []string{
+				"mkdir", "-p", dirname,
+			},
+			User: specs.User{
+				UID: 0,
+				GID: 0,
+			},
+			Env: []string{
+				"PATH=/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin",
+			},
+			NoNewPrivileges: false,
+		},
+	}
+	name := stringid.GenerateRandomID()[:20]
+
+	testerSpec.Mounts = []specs.Mount{
+		{
+			Options: []string{
+				"rbind",
+				"rw",
+				"rshared",
+			},
+			Source:      dirname,
+			Destination: dirname,
+			Type:        "bind",
+		},
+	}
+
+	cOpts := []containerd.NewContainerOpts{
+		containerd.WithNewSnapshot(name, image),
+		containerd.WithImage(image),
+		containerd.WithSpec(&testerSpec,
+			WithHostMounts,
+			WithAllCapabilities,
+			WithHostPropagation,
+		),
+	}
+
+	// Start by trying to mount it - if we can create the container, it exists and we're done
+
+	container, err := client.NewContainer(
+		ctx,
+		name,
+		cOpts...,
+	)
+	if err != nil {
+		return fmt.Errorf("Failed to create container to evaluate host dir %s: %s", dirname, err)
+	}
+	task, err := container.NewTask(ctx, cio.NullIO)
+	if err == nil {
+		// Path exists, kill the task and proceed to clean up
+		err = task.Kill(ctx, SIGKILL)
+		if err != nil && !strings.Contains(err.Error(), "not found") {
+			return fmt.Errorf("failed to kill task: %s", err)
+		}
+		exitStatus, err := task.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("Failed to wait for container task: %s", err)
+		}
+		<-exitStatus
```

Comment:
Perhaps, but the only thing that matters is the task was created, since mount validation is done at task creation time.  If the path doesn't exist, the task create fails.